README
NAME
    Dist::Zilla::Plugin::Twitter - Twitter when you release with Dist::Zilla

VERSION
    version 0.025

SYNOPSIS
    In your dist.ini:

        [Twitter]
        hash_tags = #foo
        url_shortener = TinyURL

DESCRIPTION
    This plugin will use Net::Twitter to send a release notice to Twitter.
    By default, it will include a link to release on <http://metacpan.org>.

    The default configuration is as follows:

        [Twitter]
        tweet_url = https://metacpan.org/release/{{$AUTHOR_UC}}/{{$DIST}}-{{$VERSION}}/
        tweet = Released {{$DIST}}-{{$VERSION}}{{$TRIAL}} {{$URL}} !META{resources}{repository}{web}
        url_shortener = TinyURL

    The "tweet_url" is shortened with WWW::Shorten::TinyURL or whichever
    other service you choose (use 'none' to use the full URL, in which case
    Twitter will shorten it for you) and appended to the "tweet" message.

  VARIABLE INTERPOLATION
    The following variables are available for substitution in the URL and
    message templates:

          DIST        # Foo-Bar
          MODULE      # Foo::Bar
          ABSTRACT    # Foo-Bar is a module that FooBars
          VERSION     # 1.23
          TRIAL       # -TRIAL if is_trial, empty string otherwise.
          TARBALL     # Foo-Bar-1.23.tar.gz
          AUTHOR_UC   # JOHNDOE
          AUTHOR_LC   # johndoe
          AUTHOR_PATH # J/JO/JOHNDOE
          URL         # http://tinyurl.com/...

   DISTMETA INTERPOLATION
    Resources information available in the META.* files of the distribution
    can be accessed via "<META{key}{subkey}[1]">. You may mix-and-match
    "{...}" to access hashref elements and "[\d]" to access arrayref
    elements. You're responsible for making sure you are accessing the right
    part of the META data structure, and treating it as the right type of
    data. See CPAN::Meta::Spec and the "$TYPE DATA" sections of CPAN::Meta
    in particular.

    The "META{...}" replacement may also have one of two modifiers, which
    are prefixed directly before "META":

    "!" - URL shortening
        Providing an exclamation point ("!META{...}") will URL-shorten the
        value you extract from the distmeta data structure. This will have
        no effect unless the value is a URL to begin with.

    "@" - Arrayref stringification
        Providing an at-symbol (@META{...}) will include all the elements of
        the arrayref you specify by joining them with $". So, this is just
        like doing "@{ $your_array_ref }".

    So, for example, to use the GitHub home of the project instead of its
    metacpan page, one can do:

        [Twitter]
        tweet = Released {{$DIST}}-{{$VERSION}}{{$TRIAL}} !META{resource}{repository}{web}
        url_shortener = TinyURL

    Or, to include the authors in your tweet:

        [Twitter]
        tweet = @META{author} released {{$MODULE}} {{$VERSION}}: {{$URL}}

  PAUSEID
    You must be using the "UploadToCPAN" or "FakeRelease" plugin for this
    plugin to determine your PAUSEID.

  HASHTAGS
    You can use the "hash_tags" option to append hash tags (or anything,
    really) to the end of the message generated from "tweet".

        [Twitter]
        hash_tags = #perl #cpan #foo

AVAILABILITY
    The project homepage is
    <http://metacpan.org/release/Dist-Zilla-Plugin-Twitter/>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <https://metacpan.org/module/Dist::Zilla::Plugin::Twitter/>.

SOURCE
    The development version is on github at
    <http://github.com/doherty/Dist-Zilla-Plugin-Twitter> and may be cloned
    from <git://github.com/doherty/Dist-Zilla-Plugin-Twitter.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at
    <https://github.com/doherty/Dist-Zilla-Plugin-Twitter/issues>.

AUTHORS
    *   David Golden <dagolden@cpan.org>

    *   Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Dist-Zilla-Plugin-Twitter

0.025     2014-01-16
  - Twitter API is now SSL-only, so use SSL to connect to Twitter

0.024     2013-11-27
  - Add "=head1 NAME" so the pod will get munged during test

0.023     2013-10-18
  - Improved compile test

0.022     2013-09-27
  - Add new ways of interpolating metadata into tweet text, and
    use that to fix the failing tests.

0.021     2013-06-13
  - API v1.1 support [GH #5]

0.020     2013-02-02
  - Link to doherty's github repo instead of dagolden's

0.019     2013-02-02
  - Fix a typo in the default tweet spec

0.018     2013-02-02
  - Allow adding any information from META [Yanick Champoux] and
    add a link to the repo's web URL by default

0.017     2012-07-06
  - Use url_shortener=none to use the full URL (which Twitter will
    shorten with t.co)

0.016     2012-05-18
  - Don't treat the PIN as a password [GH #1]

0.015     2012-04-01 22:40:05 EST5EDT
  - Update Twitter authentication method [RT #75995 - Mike Doherty]
    - This will require all users to re-authorize the application
      at Twitter, and will fetch & store an access token + secret
      pair instead of using passwords. You may delete the section
      in your .netrc file for api.twitter.com unless you need it
      for something else.

0.014     2012-01-13 12:36:58 EST5EDT
  - distribution abstract now available for use in tweets [Harley Pig]

0.013     2011-12-01 09:06:37 EST5EDT
  - changed default URL to metacpan [Mike Doherty]

0.012     2011-11-22 17:04:03 EST5EDT
  - better handling of missing .netrc auth data [William Orr]

0.011     2011-10-07 16:11:01 EST5EDT
  - add support for -TRIAL dists [Kent Fredric]

0.010     2011-01-23 15:53:51 EST5EDT
  - add 'url_shortener' option [Mike Doherty]

0.009     2010-07-27 11:32:57 EST5EDT
  - bump Dist::Zilla prereq to 4

0.008     2010-07-26 20:14:54 EST5EDT
  - document that FakeRelease may also be used [contributed by kent fredric]
  - add some alternate sources for username since dist-zilla
    may have renamed what gets defined by release plugins
  - update dist.ini to @DAGOLDEN bundle (but still twitter it)

0.007     2010-04-29 10:09:15 EST5EDT
  - add 'hash_tags' option to append custom tags [contributed by mschout]

0.006     2010-04-28 13:10:03 EST5EDT
  - use xauth authentication

0.005     2010-04-27 10:06:14 EST5EDT
  - turn on auto collection of prerequisites (duh)

0.004     2010-04-26 17:12:38 EST5EDT
  - don't index sample distributin for testing

0.003     2010-04-26 00:26:26 EST5EDT
  - no changes -- releasing to test the new URL

0.002     2010-04-26 00:19:42 EST5EDT
  - change url from frepan to the README on cpan.cpantesters.org;
    this is one of the fastest hyperlinks to be available after upload

0.001     2010-04-25 23:46:15 EST5EDT
  - first draft


t/lib/Net/Netrc.pm
#
# This file is part of Dist-Zilla-Plugin-Twitter
#
# This software is Copyright (c) 2011 by David Golden.
#
# This is free software, licensed under:
#
#   The Apache License, Version 2.0, January 2004
#
# mock version
use strict;
use warnings;
package Net::Netrc;

my $fake = {
  login => 'jdoe@example.com',
  account => 'jdoe',
  password => 'example',
};

sub lookup { return bless $fake }

sub login { return $fake->{login} }
sub account { return $fake->{account} }
sub password { return $fake->{password} }
sub lpa { ($fake->login, $fake->password, $fake->account) }

1;


t/lib/Test/DZil.pm
#
# This file is part of Dist-Zilla-Plugin-Twitter
#
# This software is Copyright (c) 2011 by David Golden.
#
# This is free software, licensed under:
#
#   The Apache License, Version 2.0, January 2004
#
use strict;
use warnings;
package Test::DZil;

use Dist::Zilla::Tester;
use Params::Util qw(_HASH0);

use Sub::Exporter -setup => {
  exports => [
    dist_ini   => \'_dist_ini',
    simple_ini => \'_simple_ini',
  ],
  groups  => [ default => [ qw(dist_ini simple_ini) ] ],
};

sub _build_ini_builder {
  my ($starting_core) = @_;
  $starting_core ||= {};

  sub {
    my (@arg) = @_;
    my $new_core = _HASH0($arg[0]) ? shift(@arg) : {};

    my $core_config = { %$starting_core, %$new_core };

    my $config = '';

    for my $key (keys %$core_config) {
      my @values = ref $core_config->{ $key }
                 ? @{ $core_config->{ $key } }
                 : $core_config->{ $key };

      $config .= "$key = $_\n" for grep {defined} @values;
    }

    $config .= "\n" if length $config;

    for my $line (@arg) {
      my @plugin = ref $line ? @$line : ($line, {});
      my $moniker = shift @plugin;
      my $name    = _HASH0($plugin[0]) ? undef : shift @plugin;
      my $payload = shift(@plugin) || {};

      die "TOO MANY ARGS TO PLUGIN GAHLGHALAGH" if @plugin;

      $config .= '[' . $moniker;
      $config .= ' / ' . $name if defined $name;
      $config .= "]\n";

      for my $key (keys %$payload) {
        my @values = ref $payload->{ $key }
                   ? @{ $payload->{ $key } }
                   : $payload->{ $key };

        $config .= "$key = $_\n" for @values;
      }

      $config .= "\n";
    }

    return $config;
  }
}

sub _dist_ini {
  _build_ini_builder;
}

sub _simple_ini {
  _build_ini_builder({
    name     => 'DZT-Sample',
    abstract => 'Sample DZ Dist',
    version  => '0.001',
    author   => 'E. Xavier Ample <example@example.org>',
    license  => 'Perl_5',
    copyright_holder => 'E. Xavier Ample',
  });
}

1;


t/lib/LWP/TestUA.pm
#
# This file is part of Dist-Zilla-Plugin-Twitter
#
# This software is Copyright (c) 2011 by David Golden.
#
# This is free software, licensed under:
#
#   The Apache License, Version 2.0, January 2004
#
# crudly adapted from t/lib/TestUA.pm in Net::Twitter
use strict;
use warnings;
use HTTP::Response;
package LWP::TestUA;

use base 'LWP::UserAgent';

# from http://apiwiki.twitter.com/ xAuth example
my $token_reply =
  "oauth_token=819797-torCkTs0XK7H2A2i1ee5iofqkMC4p7aayeEXRTmlw&" .
  "oauth_token_secret=SpuaLXRxZ0gOZHNQKPooBiWC2RY81klw13kLZGa2wc&" .
  "user_id=819797&screen_name=episod";

sub new {
  my $class = shift;
  my $self = $class->SUPER::new(@_);
  $self->add_handler(
    request_send => sub {
      my $req = shift;
      my $res = HTTP::Response->new(200, 'OK');
      if ( $req->uri =~ qr{\Ahttps://api.twitter.com/oauth/access_token} ) {
        $res->content($token_reply);
      }
      else {
        $res->content('{"test":"success"}');
      }
      return $res
    },
  );
  return $self;
}

1;



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Twitter when you release with Dist::Zilla",
  "dist_author" => [
    "David Golden <dagolden\@cpan.org>",
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-Twitter",
  "dist_version" => "0.025",
  "license" => "apache",
  "module_name" => "Dist::Zilla::Plugin::Twitter",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Config::INI::Reader" => 0,
    "Config::INI::Writer" => 0,
    "Dist::Zilla" => 4,
    "Dist::Zilla::Role::AfterRelease" => 0,
    "Dist::Zilla::Role::TextTemplate" => 0,
    "Dist::Zilla::Util" => 0,
    "File::Path" => 0,
    "File::Spec" => 0,
    "Moose" => "0.99",
    "Net::Twitter" => "4.00001",
    "Try::Tiny" => 0,
    "WWW::Shorten" => "3.02",
    "WWW::Shorten::Simple" => 0,
    "WWW::Shorten::TinyURL" => 0,
    "namespace::autoclean" => "0.09",
    "perl" => "5.009005",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dist::Zilla::App::Tester" => 0,
    "Dist::Zilla::Role::Releaser" => 0,
    "Dist::Zilla::Tester" => 0,
    "File::Spec" => 0,
    "HTTP::Response" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "LWP::UserAgent" => 0,
    "Params::Util" => 0,
    "Sub::Exporter" => 0,
    "Test::More" => "0.88",
    "base" => 0,
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "Dist::Zilla::App::Tester" => 0,
  "Dist::Zilla::Role::Releaser" => 0,
  "Dist::Zilla::Tester" => 0,
  "File::Spec" => 0,
  "HTTP::Response" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "LWP::UserAgent" => 0,
  "Module::Build" => "0.3601",
  "Params::Util" => 0,
  "Sub::Exporter" => 0,
  "Test::More" => "0.88",
  "base" => 0,
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.009005;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Twitter when you release with Dist::Zilla",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>, Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Twitter",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Dist::Zilla::Plugin::Twitter",
  "PREREQ_PM" => {
    "Config::INI::Reader" => 0,
    "Config::INI::Writer" => 0,
    "Dist::Zilla" => 4,
    "Dist::Zilla::Role::AfterRelease" => 0,
    "Dist::Zilla::Role::TextTemplate" => 0,
    "Dist::Zilla::Util" => 0,
    "File::Path" => 0,
    "File::Spec" => 0,
    "Moose" => "0.99",
    "Net::Twitter" => "4.00001",
    "Try::Tiny" => 0,
    "WWW::Shorten" => "3.02",
    "WWW::Shorten::Simple" => 0,
    "WWW::Shorten::TinyURL" => 0,
    "namespace::autoclean" => "0.09",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::App::Tester" => 0,
    "Dist::Zilla::Role::Releaser" => 0,
    "Dist::Zilla::Tester" => 0,
    "File::Spec" => 0,
    "HTTP::Response" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "LWP::UserAgent" => 0,
    "Params::Util" => 0,
    "Sub::Exporter" => 0,
    "Test::More" => "0.88",
    "base" => 0,
    "lib" => 0
  },
  "VERSION" => "0.025",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Config::INI::Reader" => 0,
  "Config::INI::Writer" => 0,
  "Dist::Zilla" => 4,
  "Dist::Zilla::App::Tester" => 0,
  "Dist::Zilla::Role::AfterRelease" => 0,
  "Dist::Zilla::Role::Releaser" => 0,
  "Dist::Zilla::Role::TextTemplate" => 0,
  "Dist::Zilla::Tester" => 0,
  "Dist::Zilla::Util" => 0,
  "File::Path" => 0,
  "File::Spec" => 0,
  "HTTP::Response" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "LWP::UserAgent" => 0,
  "Module::Build" => "0.3601",
  "Moose" => "0.99",
  "Net::Twitter" => "4.00001",
  "Params::Util" => 0,
  "Sub::Exporter" => 0,
  "Test::More" => "0.88",
  "Try::Tiny" => 0,
  "WWW::Shorten" => "3.02",
  "WWW::Shorten::Simple" => 0,
  "WWW::Shorten::TinyURL" => 0,
  "base" => 0,
  "lib" => 0,
  "namespace::autoclean" => "0.09",
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



